src/databricks/labs/ucx/assessment/crawlers.py

```diff
@@ -80,6 +94,47 @@ def spark_version_compatibility(spark_version: str) -> str:
     return "supported"
 
 
+class AzureServicePrincipalCrawler(CrawlerBase):
+    def __init__(self, _azure_spn_list_with_data_access: list, ws: WorkspaceClient, sbe: SqlBackend, schema):
+        super().__init__(sbe, "hive_metastore", schema, "azure_service_principals")
+        self._ws = ws
+        self._azure_spn_list_with_data_access = _azure_spn_list_with_data_access
+
+    def _crawl(self) -> list[AzureServicePrincipalInfo]:
+        all_service_principals = list(self._ws.service_principals.list())
+        return list(self._assess_service_principals(all_service_principals))
+
+    def _list_all_cluster_with_spn_in_spark_conf(self):
+        for cluster in self._ws.clusters.list():
+            if cluster.cluster_source != ClusterSource.JOB:
+                if cluster.spark_conf is not None:
+                    if _azure_sp_conf_present_check(cluster.spark_conf):
+                        spn_application_id = _get_azure_spn_application_id(cluster.spark_conf)
+                        if spn_application_id:
+                            matched = re.search(_SECRET_PATTERN, spn_application_id)
+                            if matched:
+                                spn_application_id = self._ws.secrets.get_secret(
+                                    matched.group(1).split("/")[1], matched.group(1).split("/")[2]
+                                )
+                            self._azure_spn_list_with_data_access.append(spn_application_id)
```

Comment:
this code is difficult to read. it can be simplified if you rewrite this with `if not ...: continue` way.

Reply:
also, this code wout not detect if spn_application_id is not in secret scope.

Reply:
I tested with` if not...continue` but there are situations like - cluster spark conf needs to checked as well as policy spark conf. If using continue and cluster spark conf not present the control will move to the next value in loop w/o checking the policy conf. If I am missing something we can cover that during the call.